File: app/src/main/java/fr/insapp/insapp/activities/PostActivity.java
package fr.insapp.insapp.activities;

import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.util.Linkify;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.ContentViewEvent;
import com.like.LikeButton;
import com.like.OnLikeListener;

import de.hdodenhof.circleimageview.CircleImageView;
import fr.insapp.insapp.R;
import fr.insapp.insapp.adapters.CommentRecyclerViewAdapter;
import fr.insapp.insapp.http.ServiceGenerator;
import fr.insapp.insapp.listeners.PostCommentLongClickListener;
import fr.insapp.insapp.models.Club;
import fr.insapp.insapp.models.Notification;
import fr.insapp.insapp.models.Notifications;
import fr.insapp.insapp.models.Post;
import fr.insapp.insapp.models.PostInteraction;
import fr.insapp.insapp.models.User;
import fr.insapp.insapp.utility.CommentEditText;
import fr.insapp.insapp.utility.RatioImageView;
import fr.insapp.insapp.utility.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;

/**
 * Created by thomas on 12/11/2016.
 */

public class PostActivity extends AppCompatActivity {

    private CommentRecyclerViewAdapter adapter;

    private Post post;
    private Club club;

    // post

    private CircleImageView clubAvatarCircleImageView;
    private TextView titleTextView;
    private LikeButton likeButton;
    private TextView likeCounterTextView;
    private TextView descriptionTextView;
    private TextView dateTextView;
    private CircleImageView userAvatarCircleImageView;
    private ImageView imageView;
    private RatioImageView placeholderImageView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_post);

        // building ui

        this.clubAvatarCircleImageView = (CircleImageView) findViewById(R.id.post_club_avatar);
        this.titleTextView = (TextView) findViewById(R.id.post_title);
        this.likeButton = (LikeButton) findViewById(R.id.post_like_button);
        this.likeCounterTextView = (TextView) findViewById(R.id.post_like_counter);
        this.descriptionTextView = (TextView) findViewById(R.id.post_text);
        this.dateTextView = (TextView) findViewById(R.id.post_date);
        this.userAvatarCircleImageView = (CircleImageView) findViewById(R.id.comment_post_username_avatar);
        this.imageView = (ImageView) findViewById(R.id.post_image);
        this.placeholderImageView = (RatioImageView) findViewById(R.id.post_placeholder);

        // post

        Intent intent = getIntent();
        this.post = intent.getParcelableExtra("post");

        final User user = Utils.getUser();

        // Answers

        Answers.getInstance().logContentView(new ContentViewEvent()
                .putContentId(post.getId())
                .putContentName(post.getTitle())
                .putContentType("Post")
                .putCustomAttribute("Favorites count", post.getLikes().size())
                .putCustomAttribute("Comments count", post.getComments().size()));

        // hide image if necessary

        if (post.getImageSize() == null || post.getImage().isEmpty()) {
            placeholderImageView.setVisibility(View.GONE);
            imageView.setVisibility(View.GONE);
        }

        // mark notification as seen

        if (intent.getParcelableExtra("notification") != null) {
            final Notification notification = intent.getParcelableExtra("notification");

            Call<Notifications> call = ServiceGenerator.create().markNotificationAsSeen(user.getId(), notification.getId());
            call.enqueue(new Callback<Notifications>() {
                @Override
                public void onResponse(@NonNull Call<Notifications> call, @NonNull Response<Notifications> response) {
                    if (!response.isSuccessful()) {
                        Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Notifications> call, @NonNull Throwable t) {
                    Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                }
            });
        }

        // toolbar

        Toolbar toolbar = (Toolbar) findViewById(R.id.post_toolbar);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                final Drawable upArrow = ContextCompat.getDrawable(PostActivity.this, R.drawable.abc_ic_ab_back_material);
                upArrow.setColorFilter(0xffffffff, PorterDuff.Mode.SRC_ATOP);
                getSupportActionBar().setHomeAsUpIndicator(upArrow);
            }
        }

        // like button

        likeButton.setLiked(post.isPostLikedBy(user.getId()));
        likeCounterTextView.setText(post.getLikes().size() + "");

        likeButton.setOnLikeListener(new OnLikeListener() {
             @Override
             public void liked(LikeButton likeButton) {
                 likeCounterTextView.setText(Integer.valueOf((String) likeCounterTextView.getText()) + 1 + "");

                 Call<PostInteraction> call = ServiceGenerator.create().likePost(post.getId(), user.getId());
                 call.enqueue(new Callback<PostInteraction>() {
                     @Override
                     public void onResponse(@NonNull Call<PostInteraction> call, @NonNull Response<PostInteraction> response) {
                         if (response.isSuccessful()) {
                             post = response.body().getPost();
                         }
                         else {
                             Toast.makeText(PostActivity.this, "PostRecyclerViewAdapter", Toast.LENGTH_LONG).show();
                         }
                     }

                     @Override
                     public void onFailure(@NonNull Call<PostInteraction> call, @NonNull Throwable t) {
                         Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                     }
                 });
             }

             @Override
             public void unLiked(LikeButton likeButton) {
                 likeCounterTextView.setText(Integer.valueOf((String) likeCounterTextView.getText()) - 1 + "");

                 if (Integer.valueOf((String) likeCounterTextView.getText()) < 0) {
                     likeCounterTextView.setText("0");
                 }

                 Call<PostInteraction> call = ServiceGenerator.create().dislikePost(post.getId(), user.getId());
                 call.enqueue(new Callback<PostInteraction>() {
                     @Override
                     public void onResponse(@NonNull Call<PostInteraction> call, @NonNull Response<PostInteraction> response) {
                         if (response.isSuccessful()) {
                             post = response.body().getPost();
                         }
                         else {
                             Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                         }
                     }

                     @Override
                     public void onFailure(@NonNull Call<PostInteraction> call, @NonNull Throwable t) {
                         Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                     }
                 });
             }
        });

        generateActivity();
    }

    private void generateActivity() {
        Call<Club> call = ServiceGenerator.create().getClubFromId(post.getAssociation());
        call.enqueue(new Callback<Club>() {
            @Override
            public void onResponse(@NonNull Call<Club> call, @NonNull Response<Club> response) {
                if (response.isSuccessful()) {
                    club = response.body();

                    Glide
                            .with(PostActivity.this)
                            .load(ServiceGenerator.CDN_URL + club.getProfilePicture())
                            .transition(withCrossFade())
                            .into(clubAvatarCircleImageView);

                    // listener

                    clubAvatarCircleImageView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            startActivity(new Intent(PostActivity.this, ClubActivity.class).putExtra("club", club));
                        }
                    });
                }
                else {
                    Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(@NonNull Call<Club> call, @NonNull Throwable t) {
                Toast.makeText(PostActivity.this, "PostActivity", Toast.LENGTH_LONG).show();
            }
        });

        this.titleTextView.setText(post.getTitle());
        this.descriptionTextView.setText(post.getDescription());
        this.dateTextView.setText(Utils.displayedDate(post.getDate()));

        // view links contained in description

        Linkify.addLinks(descriptionTextView, Linkify.ALL);
        Utils.convertToLinkSpan(PostActivity.this, descriptionTextView);

        // adapter

        this.adapter = new CommentRecyclerViewAdapter(PostActivity.this, Glide.with(this), post.getComments());
        adapter.setOnItemLongClickListener(new PostCommentLongClickListener(PostActivity.this, post, adapter));

        // edit comment

        CommentEditText commentEditText = (CommentEditText) findViewById(R.id.comment_post_input);
        commentEditText.setupComponent(adapter, post);

        // recycler view

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerview_comments_post);
        recyclerView.setHasFixedSize(true);
        recyclerView.setNestedScrollingEnabled(false);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);

        recyclerView.setAdapter(adapter);

        // retrieve the avatar of the user

        final User user = Utils.getUser();

        final int id = getResources().getIdentifier(Utils.drawableProfileName(user.getPromotion(), user.getGender()), "drawable", getPackageName());
        Glide
                .with(PostActivity.this)
                .load(id)
                .transition(withCrossFade())
                .into(userAvatarCircleImageView);

        // image

        if (post.getImageSize() != null && !post.getImage().isEmpty()) {
            placeholderImageView.setImageSize(post.getImageSize());

            Glide
                    .with(PostActivity.this)
                    .load(ServiceGenerator.CDN_URL + post.getImage())
                    .transition(withCrossFade())
                    .into(imageView);
        }
    }

    @Override
    public void finish() {
        Intent sendIntent = new Intent();
        sendIntent.putExtra("post", post);

        setResult(RESULT_OK, sendIntent);

        super.finish();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isTaskRoot()) {
                    startActivity(new Intent(PostActivity.this, MainActivity.class));
                }
                else {
                    finish();
                }
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/fr/insapp/insapp/activities/PostActivity.kt
package fr.insapp.insapp.activities

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.text.util.Linkify
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade
import com.bumptech.glide.request.RequestOptions
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.ContentViewEvent
import com.varunest.sparkbutton.SparkEventListener
import fr.insapp.insapp.R
import fr.insapp.insapp.adapters.CommentRecyclerViewAdapter
import fr.insapp.insapp.http.ServiceGenerator
import fr.insapp.insapp.models.*
import fr.insapp.insapp.utility.Utils
import kotlinx.android.synthetic.main.activity_post.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Created by thomas on 12/11/2016.
 */

class PostActivity : AppCompatActivity() {

    private lateinit var commentAdapter: CommentRecyclerViewAdapter

    private lateinit var post: Post
    private var association: Association? = null

    private lateinit var requestManager: RequestManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestManager = Glide.with(this)

        val user = Utils.user

        // post

        if (intent.getParcelableExtra<Post>("post") != null) {
            // coming from navigation

            this.post = intent.getParcelableExtra("post")
            generateActivity()

            // mark notification as seen

            if (intent.getParcelableExtra<Notification>("notification") != null) {
                val notification = intent.getParcelableExtra<Notification>("notification")

                if (user != null) {
                    val call = ServiceGenerator.client.markNotificationAsSeen(user.id, notification.id)
                    call.enqueue(object : Callback<Notifications> {
                        override fun onResponse(call: Call<Notifications>, response: Response<Notifications>) {
                            if (!response.isSuccessful) {
                                Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                            }
                        }

                        override fun onFailure(call: Call<Notifications>, t: Throwable) {
                            Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                        }
                    })
                } else {
                    Log.d(TAG, "")
                }
            }
        } else {
            // coming from notification
            setContentView(R.layout.loading)

            ServiceGenerator.client.getPostFromId(intent.getStringExtra("ID"))
                .enqueue(object : Callback<Post> {
                    override fun onResponse(call: Call<Post>, response: Response<Post>) {
                        val result = response.body()
                        if (response.isSuccessful && result != null) {
                            post = result
                            generateActivity()
                        } else {
                            Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                        }
                    }

                    override fun onFailure(call: Call<Post>, t: Throwable) {
                        Toast.makeText(this@PostActivity, R.string.check_internet_connection, Toast.LENGTH_LONG).show()

                        // Open the application
                        startActivity(Intent(this@PostActivity, MainActivity::class.java))
                        finish()
                    }
                })
        }
    }

    private fun generateActivity() {
        setContentView(R.layout.activity_post)

        // toolbar

        setSupportActionBar(post_toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val user = Utils.user

        // Answers

        Answers.getInstance().logContentView(ContentViewEvent()
            .putContentId(post.id)
            .putContentName(post.title)
            .putContentType("Post")
            .putCustomAttribute("Favorites count", post.likes.size)
            .putCustomAttribute("Comments count", post.comments.size))

        // hide image if necessary

        if (post.image.isEmpty()) {
            post_placeholder?.visibility = View.GONE
            post_image?.visibility = View.GONE
        }

        // like button

        user?.let {
            post_like_button?.setChecked(post.isPostLikedBy(user.id))
        }
        post_like_counter?.text = post.likes.size.toString()

        post_like_button?.setEventListener(object : SparkEventListener {
            override fun onEvent(icon: ImageView, buttonState: Boolean) {
                if (buttonState) {
                    post_like_counter?.text = (Integer.valueOf(post_like_counter?.text as String) + 1).toString()

                    if (user != null) {
                        val call = ServiceGenerator.client.likePost(post.id, user.id)
                        call.enqueue(object : Callback<PostInteraction> {
                            override fun onResponse(call: Call<PostInteraction>, response: Response<PostInteraction>) {
                                val results = response.body()
                                if (response.isSuccessful && results != null) {
                                    post = results.post
                                } else {
                                    Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                                }
                            }

                            override fun onFailure(call: Call<PostInteraction>, t: Throwable) {
                                Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                            }
                        })
                    } else {
                        Log.d(TAG, "Couldn't update the like status: user is null")
                    }
                } else {
                    post_like_counter?.text = (Integer.valueOf(post_like_counter?.text as String) - 1).toString()

                    if (Integer.valueOf(post_like_counter?.text as String) < 0) {
                        post_like_counter?.text = "0"
                    }

                    if (user != null) {
                        val call = ServiceGenerator.client.dislikePost(post.id, user.id)
                        call.enqueue(object : Callback<PostInteraction> {
                            override fun onResponse(call: Call<PostInteraction>, response: Response<PostInteraction>) {
                                val result = response.body()
                                if (response.isSuccessful && result != null) {
                                    post = result.post
                                } else {
                                    Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                                }
                            }

                            override fun onFailure(call: Call<PostInteraction>, t: Throwable) {
                                Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                            }
                        })
                    } else {
                        Log.d(TAG, "Couldn't update the like status: user is null")
                    }
                }
            }

            override fun onEventAnimationEnd(button: ImageView?, buttonState: Boolean) {}

            override fun onEventAnimationStart(button: ImageView?, buttonState: Boolean) {}
        })

        val call = ServiceGenerator.client.getAssociationFromId(post.association)
        call.enqueue(object : Callback<Association> {
            override fun onResponse(call: Call<Association>, response: Response<Association>) {
                if (response.isSuccessful) {
                    association = response.body()

                    requestManager
                        .load(ServiceGenerator.CDN_URL + association!!.profilePicture)
                        .apply(RequestOptions.circleCropTransform())
                        .transition(withCrossFade())
                        .into(post_association_avatar)

                    // listener

                    post_association_avatar?.setOnClickListener { startActivity(Intent(this@PostActivity, AssociationActivity::class.java).putExtra("association", association)) }
                } else {
                    Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<Association>, t: Throwable) {
                Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
            }
        })

        post_title?.text = post.title
        post_text?.text = post.description
        post_date?.text = Utils.displayedDate(post.date)

        // view links contained in description

        Linkify.addLinks(post_text, Linkify.ALL)
        Utils.convertToLinkSpan(this@PostActivity, post_text)

        // edit comment adapter

        commentAdapter = CommentRecyclerViewAdapter(post.comments, requestManager)
        commentAdapter.setOnCommentItemLongClickListener(PostCommentLongClickListener(this@PostActivity, post, commentAdapter))

        comment_post_input.setupComponent()
        comment_post_input.setOnEditorActionListener(TextView.OnEditorActionListener { textView, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                // hide keyboard

                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(textView.windowToken, 0)

                val content = comment_post_input.text.toString()
                comment_post_input.text.clear()

                if (content.isNotBlank()) {
                    user?.let {
                        val comment = Comment(null, user.id, content, null, comment_post_input.tags)

                        Log.d(TAG, comment.toString())

                        ServiceGenerator.client.commentPost(post.id, comment)
                            .enqueue(object : Callback<Post> {
                                override fun onResponse(call: Call<Post>, response: Response<Post>) {
                                    if (response.isSuccessful) {
                                        commentAdapter.addComment(comment)
                                        comment_post_input.tags.clear()

                                        Toast.makeText(this@PostActivity, resources.getText(R.string.write_comment_success), Toast.LENGTH_LONG).show()
                                    } else {
                                        Log.e(TAG, "Couldn't post comment: ${response.code()}")
                                        Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                                    }
                                }

                                override fun onFailure(call: Call<Post>, t: Throwable) {
                                    Log.e(TAG, "Couldn't post comment: ${t.message}")
                                    Toast.makeText(this@PostActivity, TAG, Toast.LENGTH_LONG).show()
                                }
                            })
                    }
                }

                return@OnEditorActionListener true
            }

            false
        })

        // recycler view

        recyclerview_comments_post.setHasFixedSize(true)
        recyclerview_comments_post.isNestedScrollingEnabled = false

        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerview_comments_post.layoutManager = layoutManager

        recyclerview_comments_post.adapter = commentAdapter

        // retrieve the avatar of the user

        val id = resources.getIdentifier(Utils.drawableProfileName(user?.promotion, user?.gender), "drawable", packageName)
        requestManager
            .load(id)
            .transition(withCrossFade())
            .apply(RequestOptions.circleCropTransform())
            .into(comment_post_username_avatar)

        // image

        if (post.image.isNotEmpty()) {
            post_placeholder?.setImageSize(post.imageSize)

            requestManager
                .load(ServiceGenerator.CDN_URL + post.image)
                .transition(withCrossFade())
                .into(post_image)
        }
    }

    override fun finish() {
        if (::post.isInitialized) {
            val sendIntent = Intent()
            sendIntent.putExtra("post", post)
            setResult(Activity.RESULT_OK, sendIntent)
        } else {
            setResult(Activity.RESULT_CANCELED)
        }

        super.finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                if (isTaskRoot) {
                    startActivity(Intent(this@PostActivity, MainActivity::class.java))
                } else {
                    finish()
                }
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    companion object {

        const val TAG = "PostActivity"
    }

    class PostCommentLongClickListener(
        private val context: Context,
        private val post: Post,
        private val adapter: CommentRecyclerViewAdapter
    ): CommentRecyclerViewAdapter.OnCommentItemLongClickListener {

        override fun onCommentItemLongClick(comment: Comment) {
            val alertDialogBuilder = AlertDialog.Builder(context)
            val user = Utils.user

            // delete comment

            if (user != null) {
                if (user.id == comment.user) {
                    alertDialogBuilder
                        .setTitle(context.resources.getString(R.string.delete_comment_action))
                        .setMessage(R.string.delete_comment_are_you_sure)
                        .setCancelable(true)
                        .setPositiveButton(context.getString(R.string.positive_button)) { _: DialogInterface, _: Int ->
                            val call = ServiceGenerator.client.uncommentPost(post.id, comment.id!!)
                            call.enqueue(object : Callback<Post> {
                                override fun onResponse(call: Call<Post>, response: Response<Post>) {
                                    val post = response.body()
                                    if (response.isSuccessful && post != null) {
                                        adapter.setComments(post.comments)
                                    } else {
                                        Toast.makeText(context, "PostCommentLongClickListener", Toast.LENGTH_LONG).show()
                                    }
                                }

                                override fun onFailure(call: Call<Post>, t: Throwable) {
                                    Toast.makeText(context, "PostCommentLongClickListener", Toast.LENGTH_LONG).show()
                                }
                            })
                        }
                        .setNegativeButton(context.getString(R.string.negative_button)) { dialogAlert: DialogInterface, _: Int ->
                            dialogAlert.cancel()
                        }
                        .create().show()
                }

                // report comment

                else {
                    alertDialogBuilder
                        .setTitle(context.getString(R.string.report_comment_action))
                        .setMessage(R.string.report_comment_are_you_sure)
                        .setCancelable(true)
                        .setPositiveButton(context.getString(R.string.positive_button)) { _: DialogInterface, _: Int ->
                            val call = ServiceGenerator.client.reportComment(post.id, comment.id!!)
                            call.enqueue(object : Callback <Void> {
                                override fun onResponse(call: Call<Void>, response: Response<Void>) {
                                    if (response.isSuccessful) {
                                        Toast.makeText(context, context.getString(R.string.report_comment_success), Toast.LENGTH_SHORT).show();
                                    } else {
                                        Toast.makeText(context, "PostCommentLongClickListener", Toast.LENGTH_LONG).show();
                                    }
                                }

                                override fun onFailure(call: Call<Void>, t: Throwable) {
                                    Toast.makeText(context, "PostCommentLongClickListener", Toast.LENGTH_LONG).show();
                                }
                            })
                        }
                        .setNegativeButton(context.getString(R.string.negative_button)) { dialogAlert: DialogInterface, _: Int ->
                            dialogAlert.cancel()
                        }
                        .create().show()
                }
            }
        }
    }
}

fun Post.isPostLikedBy(userID: String): Boolean {
    for (idUser in likes)
        if (idUser == userID) return true

    return false
}